### PR TITLE
feat: improve callback manager

### PR DESCRIPTION
**Pull request by himself65**
As we talked about before

Also, use AsyncLocalStorage for better API and code style.

Related: https://github.com/run-llama/LlamaIndexTS/issues/671, https://github.com/run-llama/LlamaIndexTS/pull/668#discussion_r1542166013

In the future, we should put more things into global settings and async local storage instead of using `new XXX({ ... })` (everything into parameter when creation), which worse understand and debug

Downstream: https://github.com/run-llama/LlamaIndexTS/pull/667

---

There should only be 2 ways to change such variables in the future (like callback manager, or llm in a service context)

1. AsyncLocalStorage

```ts
import { runWith } from 'llamaindex'
await runWith(myllm, () => {})
```

2. Global Settings

```ts
import { Settings } from 'llamaindex'

Settings.llm = myllm
```

---

More context about AsyncLocalStorage

- TC39 Proposal: https://github.com/tc39/proposal-async-context
- https://nodejs.org/api/async_context.html

### CODE REVIEW

File: packages/core/src/callbacks/CallbackManager.ts [hunk 1]
```diff
@@ -60,25 +67,94 @@ export interface RetrievalCallbackResponse extends BaseCallbackResponse {
...
+    handler: EventHandler<LlamaIndexEventMaps[K]>,
+  ) {
+    if (!this.#handlers.has(event)) {
+      this.#handlers.set(event, []);
+    }
+    this.#handlers.get(event)!.push(
+      // @ts-expect-error
+      handler,
+    );
+  }
```
Comment (himself65) [inline packages/core/src/callbacks/CallbackManager.ts:121]: WIP

File: examples/multimodal/rag.ts [hunk 2]
```diff
@@ -38,17 +39,18 @@ async function main() {
...
   const index = await createIndex(serviceContext);
 
   const queryEngine = index.asQueryEngine({
     responseSynthesizer: new MultiModalResponseSynthesizer({ serviceContext }),
     retriever: index.asRetriever({ similarityTopK: 3, imageSimilarityTopK: 1 }),
   });
-  const result = await queryEngine.query({
-    query: "Tell me more about Vincent van Gogh's famous paintings",
-  });
+  const result = await runWithCallbackManager(callbackManager, () =>
```
Comment (yisding) [inline examples/multimodal/rag.ts:49]: Is this runWithCallbackManager the only way to get this functionality?

I think for the most natural DevEx the user should do something like registerCallbackManager one, and then everything should just automatically flow to that.

Comment (yisding) [reply examples/multimodal/rag.ts:49]: Almost like a `window.addEventListener` (actually, that makes me wonder if that would be an acceptable interface to hook into.

Comment (himself65) [reply examples/multimodal/rag.ts:49]: As mentioned in the description, you could do `Settings.callbackManager = new CallbackManager` if you want

Comment (himself65) [reply examples/multimodal/rag.ts:49]: `runWith` is more like fine-grained tracking situation.

For example for tracking LLM cost for a single call, you call use `runWithCallbackManager(myManager, () => llm.query())`, then you could get usage for this single call.

But you want to do some global injection, I think we should suggest user to use Settings API

Comment (yisding) [reply examples/multimodal/rag.ts:49]: OK let's make that the default in our examples. I'm guessing most beginner users will only need one CallbackManager for one specific thing (i.e. logging or Weights and Biases). Btw, should we make that an array? Or is there an easy way to chain CallbackManagers?

Comment (himself65) [reply examples/multimodal/rag.ts:49]: I'm making it chainable now; see https://github.com/run-llama/LlamaIndexTS/pull/675#discussion_r1544860405

I'm imaging that you can do like 

```
cb
  .addHandler('chat-start', () => {})
  .addHandler('chat-start', () => {})
  .addHandler('chat-end', () => {})
```

File: examples/qdrantdb/preFilters.ts [hunk 3]
```diff
@@ -36,19 +37,21 @@ async function main() {
...
-        callbackManager: new CallbackManager({
-          onRetrieve: (data) => {
-            console.log(
-              "The retrieved nodes are:",
-              data.nodes.map((node) => node.node.getContent(MetadataMode.NONE)),
-            );
-          },
-        }),
+    const index = await runWithCallbackManager(
+      new CallbackManager({
```
Comment (marcusschiesser) [inline examples/qdrantdb/preFilters.ts:41]: @himself65 as we're changing this anyhow, how about simplifying it to:
```suggestion
const index = await runWithCallbacks({
```

**Pull request merged**